pkg/controllers/subnet/subnet_controller.go

```diff
@@ -0,0 +1,279 @@
+package subnet
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"os"
+	"reflect"
+	"runtime"
+
+	v1 "k8s.io/api/core/v1"
+	apimachineryruntime "k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+
+	"github.com/vmware-tanzu/nsx-operator/pkg/apis/v1alpha1"
+	"github.com/vmware-tanzu/nsx-operator/pkg/controllers/common"
+	"github.com/vmware-tanzu/nsx-operator/pkg/logger"
+	"github.com/vmware-tanzu/nsx-operator/pkg/metrics"
+	servicecommon "github.com/vmware-tanzu/nsx-operator/pkg/nsx/services/common"
+	"github.com/vmware-tanzu/nsx-operator/pkg/nsx/services/subnet"
+)
+
+var (
+	log                     = logger.Log
+	ResultNormal            = common.ResultNormal
+	ResultRequeue           = common.ResultRequeue
+	ResultRequeueAfter5mins = common.ResultRequeueAfter5mins
+	MetricResTypeSubnet     = common.MetricResTypeSubnet
+)
+
+// SubnetReconciler reconciles a SubnetSet object
+type SubnetReconciler struct {
+	Client  client.Client
+	Scheme  *apimachineryruntime.Scheme
+	Service *subnet.SubnetService
+}
+
+func (r *SubnetReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	obj := &v1alpha1.Subnet{}
+	log.Info("reconciling subnet CR", "subnet", req.NamespacedName)
+	metrics.CounterInc(r.Service.NSXConfig, metrics.ControllerSyncTotal, MetricResTypeSubnet)
+
+	if err := r.Client.Get(ctx, req.NamespacedName, obj); err != nil {
+		log.Error(err, "unable to fetch Subnet CR", "req", req.NamespacedName)
+		return ResultNormal, client.IgnoreNotFound(err)
+	}
+
+	vpcList := &v1alpha1.VPCList{}
+	if err := r.Client.List(context.Background(), vpcList, client.InNamespace(obj.Namespace)); err != nil {
+		log.Error(err, fmt.Sprintf("failed to get VPC under namespace: %s.\n", obj.Namespace))
+		return ResultRequeue, err
+	}
```

Comment:
Better to check if len(vpcList)>0 for negative case.

Reply:
Thanks, done.